src/component.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Base class for all MDL components.
+ *
+ * @export
+ */
+class MaterialComponent { // eslint-disable-line no-unused-vars
+  /**
+   * Initialize component from a DOM node.
+   *
+   * @param {Element} root The element being upgraded.
+   */
+  constructor(root) {
+    this.root_ = root;
+
+    // Check if the root has the right class.
+    if (!root.classList.contains(this.constructor.classes_.ROOT)) {
+      throw new Error(`${this.constructor.classAsString} missing ` +
+          `${this.constructor.classes_.ROOT} class.`);
+    }
+  }
+
+  // eslint-disable-next-line valid-jsdoc
+  /**
+   * CSS classes used in this component.
+   *
+   * @protected
+   * @return {Object<string, string>} The CSS classes used in this component.
+   */
+  static get classes_() {
+    // Empty in base class. Throw error if not correctly overriden.
+    throw new Error('Should have ROOT and JS keys with the style class names,' +
+      ' e.g. mdl-button and mdl-js-button.');
+  }
+
+  /**
+   * Number constants used in this component.
+   *
+   * @protected
+   * @return {Object<string, number>} The numbers used in this component.
+   */
+  static get numbers_() {
+    // Empty in base class.
+    return {};
+  }
+
+  // eslint-disable-next-line valid-jsdoc
+  /**
+   * String constants used in this component.
+   *
+   * @protected
+   * @return {Object<string, string>} The strings used in this component.
+   */
+  static get strings_() {
+    // Empty in base class. Throw error if not correctly overriden.
+    throw new Error('Should have a CLASS_NAME key with the script class name,' +
+        ' e.g. MaterialButton.');
+  }
+
+  /**
+   * Return class name as a string. Useful for automation after obfuscation.
+   *
+   * @return {string} The JS class name for this component.
+   * @suppress {missingProperties}
+   * @nocollapse
+   * @export
+   */
+  static get classAsString() {
+    return this.strings_.CLASS_NAME;
+  }
+
+  /**
+   * Initialize all self-managed instances under the given node.
+   * This will only initialize components that have specifically been marked
+   * for self-management.
+   *
+   * @param {Node} docRoot The node under which to look for components.
+   * @nocollapse
+   * @export
+   */
+  static initComponents(docRoot) {
```

Comment:
Yup, I definitely agree with the default being developer-managed lifecycles.

That said, we need some sort of mechanism for the other path (auto instantiation), and a static method on the `MaterialComponent` base class seemed like the right decision, so that each component's class inherits it as a static method as well. As we had discussed, this method won't do anything unless a signal CSS class (`material-js-foo`) is present, as per line 99.

I'm open to different approaches on how to perform auto-instantiation, and I think this bears discussion.

In the interest of getting the PR merged, though, I'll remove `initComponents` and related tests for now, so we can keep the set of features we have a consensus on and go from there.